csrc/multidevice/communication.h

```diff
@@ -34,36 +34,87 @@ enum class CommunicationType {
-  // reduced_axis is always outermost.
-  int64_t scattered_axis = -1;
-};
+using RedOpType = c10d::ReduceOp::RedOpType;
 
 // The class "Communication" represents a MPI-style communication
 // communication operation to be executed on the network. The base class
 // Communication should not be used directly but through its derived classes:
 // Broadcast, Gather, Scatter, Allgather, and SendRecv. Other collectives will
 // be added later.
+class Communication : public Expr {
+ public:
+  using Expr::Expr;
+  // Only specify `root` for types that have root.
+  // Only specify `red_op` for reduction types.
+  // Only specify `scattered_axis` for ReduceScatter.
+  //
+  // TODO: pass in input/output TV and compute root, mesh and scatteredAxis from
+  // them.
+  Communication(
+      IrBuilderPasskey passkey,
+      CommunicationType type,
+      DeviceMesh mesh,
+      DeviceIdxType root = -1,
+      RedOpType red_op = RedOpType::UNUSED,
+      int64_t scattered_axis = -1);
+
+  Communication(const Communication& other) = delete;
+  Communication& operator=(const Communication& other) = delete;
+  Communication(Communication&& other) = delete;
+  Communication& operator=(Communication&& other) = delete;
+
+  NVFUSER_DECLARE_CLONE_AND_CREATE
+
+  std::string toString(int indent_size = 0) const override;
+  std::string toInlineString(int indent_size = 0) const override;
+  const char* getOpString() const override {
+    return "Communication";
+  }
+
+  CommunicationType type() const {
+    return attribute<CommunicationType>(0);
+  }
+
+  const DeviceMesh& mesh() const {
+    return attribute<DeviceMesh>(1);
+  }
+
+  DeviceIdxType root() const {
+    return attribute<DeviceIdxType>(2);
+  }
+
+  RedOpType reduceOp() const {
+    return attribute<RedOpType>(3);
+  }
+
+  int64_t scatteredAxis() const {
+    return attribute<int64_t>(4);
+  }
 
-// CommParams contains the arguments for the communication constructors.
-// Note that each process (associated with a device index given by
-// communicator.deviceId()) will fill CommParams with different arguments,
-// depending on the role they play in this communication. For example, the root
-// of a Gather communication will have <team_size> destination buffers, whereas
-// non-root will have no destination buffers. Also, the ranks not participating
-// in the communication should not instantiate it.
+  bool isRootInMesh() const {
+    return mesh().has(root());
+  }
+
+  const Team& team();
+
+  // PyTorch's process group expects the root to be specified
+  // as an integer between 0 and world_size-1. We choose it to be
+  // the device's relative index within the team
+  int64_t getRootRelativeIndex();
+
+ private:
+  // This can be computed from params_, but given how frequently this is used
+  // in the hot path, I'm currently storing it as a field that'll be computed by
+  // Communication::team().
+  Team team_;
+};
 
 // The method "post" triggers the execution of the communication. This call is
```

Comment:
Note: I moved this comment to postSingleCommunication, because Communication::post is no longer a thing.